ToolChain/Assembler/Parser/Parser.py

```diff
@@ -20,586 +19,369 @@
-                rawLine = line
-                line = line.split()
-                if len(line) > 0:
-                    # If we make it here, there was something on the line
-                    # We move to the next step and parse the line to get the appropriate code for what's written
-                    buildInstruction, lineInstruction = self._parseCodeLine(line, rawLine)
-
-                    if buildInstruction is not None:
-                        # There was a valid instruction or memory reference on the line
-
-                        if type(buildInstruction.instructionCode) is str:
-                            # We have a memory location reference
-                            memoryReferenceDict[buildInstruction.instructionCode] = relativeAddressCounter
-                        else:
-                            # We have an instruction
-                            instructionList.append(lineInstruction)
-                            relativeAddressCounter += buildInstruction.instructionLength
-
-            except Exception as e:
-                # If all else fails, the line has invalid code
-                print("ERROR while asssembling line {}: '{}'".format(str(lineNo), line,))
-                print(e)
-                quit()
-
-        return instructionList, memoryReferenceDict, relativeAddressCounter
-
-    def _parseCodeLine(self, line: list=None, rawLine: str=None):
+    def _findInstructionCode(self, form, line, instruction):
         """
-        This will parse an extracted line of code and build an instruction from that code line
-        Parsing a line of code requires many steps.
-
-            1- Find the possible codes for the mnemonic from OperationDescription
-            2- Based on the possible codes and parts of the lines, find the correct form
-            3- Based on the form, build the instruction
-            4- Return the instruction
-                If the instruction returned has a string as instruction mnemonic, this means that the
-                instruction simply represent a label in the code. It will be removed from the final
-                program.
-
-        :param line: line split into individual arguments
-        :param rawLine: raw line as we originally read it from the file, used to preserve spaces in .dataAlpha fields
-        :return: The instruction object with all info about arguments, and the line's binary instruction
+        Finds the instruction's binary code based on its form, which relies on the arguments after the instruction.
+        Since each instruction + form pair has one distinct instruction length, we also add it here to the
+        relativeAddressCounter.
+        :param form: str, the assembled form of the particular instruction (ex: instruction-register-register)
+        :param line: str list, our line read from input split into individual arguments
+        :param instruction: bytestring, will contain our final binary code passed back to assembler
```

Comment:
From what I'm seeing in the code, the instruction parameter is not required. This comes in as an empty byte string and then is returned to the calling function. I would possibly remove it and simply create a local variable? Unless I am mistaken?

Reply:
That's definitely possible, since at this point the Instruction variable is an empty bytestring